extensions/parquet/base/src/main/java/io/deephaven/parquet/base/ColumnPageReaderImpl.java

```diff
@@ -569,25 +592,34 @@ private ValuesReader getDataReader(Encoding dataEncoding, ByteBuffer in, int val
 
     @Override
     public int numValues() throws IOException {
-        ensureNumValues();
-        return numValues;
+        if (numValues >= 0) {
+            return numValues;
+        }
+        try (final SeekableByteChannel readChannel = channelsProvider.getReadChannel(filePath)) {
+            readChannel.position(offset);
+            ensurePageHeader(readChannel);
+            // Above will automatically populate numValues
+            Assert.geq(numValues, "numValues", 0);
+            return numValues;
+        }
     }
 
     private void ensureNumValues() throws IOException {
```

Comment:
Since this method is now only called from inside ensurePageHeader, I want to move it closer to that method in the file. Will make this change as a single commit once rest of the review is done.